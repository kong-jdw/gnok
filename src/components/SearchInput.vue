<script lang="ts" setup>
import { ref } from 'vue'
import { Icon } from '@iconify/vue'

const focus = ref(false)
const field = ref()

const submit = () => {
  field.value.blur()
}

const handleKeyEvent = ({ key }: { key: string }): void => {
  if (key === 'Enter') {
    submit()
  }
}
</script>

<template>
  <div
    class="search"
    :class="{
      focus,
    }"
  >
    <Icon
      class="magnify"
      icon="material-symbols:search-rounded"
    />
    <input
      ref="field"
      placeholder="Search"
      @blur="focus = false"
      @focus="focus = true"
      @keydown="handleKeyEvent"
    >
    <Icon
      class="query"
      icon="tdesign:send-filled"
      @click="submit"
    />
  </div>
</template>

<style lang="scss" scoped>
@use '@css/media.scss';
@use '@css/colors.scss' as c;

.search {
  position: relative;

  input {
    border-radius: 4px;
    border-style: solid;
    color: c.$paragraph-light;
    outline: none;
    padding: calc(0.5rem + 1px);
    padding-left: 1.75rem;
    transition: border-color 0.1s ease-in;
    width: 100%;

    @include media.dark {
      background-color: c.$dark-gray;
      border-color: rgba(c.$bright-gray, 0.4);
      color: c.$paragraph-dark;
    }

    &:focus {
      border-color: c.$primary-blue;
    }
  }

  .iconify {
    font-size: 1rem;
    position: absolute;
    top: 50%;

    &.magnify {
      left: 1rem;
      transform: translate(-50%, -50%);
    }

    &.query {
      color: c.$primary-blue;
      cursor: pointer;
      opacity: 0.2;
      right: 1rem;
      transform: translate(50%, -50%);
      transition: opacity 0.1s ease-in;

      &:hover {
        animation: search 0.3s ease-in 1;
      }
    }
  }

  &.focus .iconify.query {
    opacity: 1;
  }
}

@keyframes search {
  0% {
    transform: translate(50%, -50%) rotate(0deg);
  }
  25% {
    transform: translate(50%, -50%) rotate(10deg);
  }
  50% {
    transform: translate(50%, -50%) rotate(0deg);
  }
  75% {
    transform: translate(50%, -50%) rotate(-10deg);
  }
  100% {
    transform: translate(50%, -50%) rotate(0deg);
  }
}
</style>
