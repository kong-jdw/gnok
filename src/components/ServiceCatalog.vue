<script lang="ts" setup>
import useServices from '@/composables/useServices'
import type { Service } from '@/composables/useServices'
import PageTitle from '@/components/PageTitle.vue'
import SearchInput from '@/components/SearchInput.vue'
import CreateServiceButton from '@/components/CreateServiceButton.vue'
import ServiceList from '@/components/ServiceList.vue'

const { services, loading }: { services: Service[], loading: boolean } = useServices()
</script>

<template>
  <div class="header-search">
    <PageTitle>Service Hub</PageTitle>
    <div class="search">
      <SearchInput />
      <CreateServiceButton />
    </div>
  </div>

  <ServiceList
    :loading="loading"
    :services="services"
  />
</template>

<style lang="scss" scoped>
@use '@css/spacing.scss';
@use '@css/media.scss';

.header-search {
  align-items: flex-start;
  display: flex;
  flex-wrap: wrap;
  gap: spacing.$space-md;
  justify-content: space-between;
  margin-bottom: spacing.$space-md;
}

.search {
  align-items: center;
  background-color: red;
  display: flex;
  gap: spacing.$space-md;
  justify-content: flex-end;
  width: 100%;

  @include media.md {
    max-width: 50%;
  }
}
</style>
