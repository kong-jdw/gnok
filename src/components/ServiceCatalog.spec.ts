import { vi, describe, it, expect, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import ServiceCatalog from './ServiceCatalog.vue'
import SearchInput from '@/components/SearchInput.vue'

const mocks = vi.hoisted(() => ({
  get: vi.fn(),
}))

vi.mock('axios', async () => {
  const actual: any = await vi.importActual('axios')
  return {
    default: {
      ...actual.default,
      // Mock get request responses
      get: mocks.get,
    },
  }
})

describe('ServiceCatalog', () => {
  beforeEach(() => {
    mocks.get.mockReset()
    setActivePinia(createPinia())
  })

  it('renders', async () => {
    const wrapper = mount(ServiceCatalog, { shallow: true })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('fetches services on mount', async () => {
    mount(ServiceCatalog, { shallow: true })

    expect(mocks.get).toHaveBeenCalledTimes(1)
    expect(mocks.get).toHaveBeenLastCalledWith('/api/services')
  })

  it('fetches new services when search changes', async () => {
    const wrapper = mount(ServiceCatalog, { shallow: true })
    expect(mocks.get).toHaveBeenCalledTimes(1)

    const search = wrapper.findComponent(SearchInput)
    await search.vm.$emit('search', 'test')
    expect(mocks.get).toHaveBeenCalledTimes(2)
    expect(mocks.get).toHaveBeenLastCalledWith('/api/services?q=test')
  })
})
