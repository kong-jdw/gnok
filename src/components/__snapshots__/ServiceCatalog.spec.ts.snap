// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceCatalog > renders 1`] = `
"<div data-v-7999d149="" class="header-search">
  <page-title-stub data-v-7999d149=""></page-title-stub>
  <div data-v-7999d149="" class="search">
    <search-input-stub data-v-7999d149=""></search-input-stub>
    <create-service-button-stub data-v-7999d149=""></create-service-button-stub>
  </div>
</div>
<service-list-stub data-v-7999d149="" services="" loading="true" error="false"></service-list-stub>"
`;
